### ISSUE TITLE
Automatic selection of sparse solver doesn't work in all cases

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#7848]
### Description

PR #7695 fixed the selection logic of the sparse solver, and was also ported to the 1.18 maintenance branch. I am currently testing 1.18.0-beta.2 for this feature, that was cherry picked into the [maintenance branch](https://github.com/OpenModelica/OpenModelica/commits/maintenance/v1.18) as well. 

If I run some of the models whose changed test results were reported in the commit (e.g. `Modelica.Electrical.Spice3.Examples.Spice3BenchmarkFourBitBinaryAdder`), I get a notification of switching to the sparse solver, and the output of `-ls=LOG_NLS_V` indeed confirms that kinsol was used instead of the default mixed solver.

However, if I try some examples from PowerGrids whereby I remove the annotation to set `-nls=kinsol`, the sparse solver is not activated.

This issue can be reproduced with this simple MWE:
```
model TestSparse
  parameter Integer N = 30;
  Real x[N];
equation
  2*x[1] + 3*x[2] = 4;
  for i in 2:N-1 loop
    1e-6*sin(x[i-1]) + 0.5*x[i] + x[i+1] = time;
  end for;
  2*x[end-1] + x[end] = 2;
  annotation(__OpenModelica_commandLineOptions = "--tearingMethod=minimalTearing",
             __OpenModelica_simulationFlags(lv="LOG_NLS_V"));
end TestSparse;
```
Now, this system has 20 unknowns, so the Jacobian has 400 elements, of which only 2+ 18*3 + 2 = 58 are nonzero, so the density should be 58/400 = 0.145, if I am not mistaken. The default threshold for [-lssMaxDensity](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-lssmaxdensity) is 0.2, so kinsol should be selected automatically here as well. Unfortunately it isn't, as the output log shows clearly. It isn't even if I set `-lssMaxDensity=0.8`.

@phannebohm could you please check what is going wrong here? I wanted to make sure that 1.18.0 would work with PowerGrids models without the need of setting the `-ls` and `-nls` flags, but it seems this still doesn't work well.

Thanks!

**Comment by phannebohm** [comment]
Looks like there is no sparsity pattern available. I'll investigate. This is urgent I guess?

**Comment by casella** [comment]
@phannebohm since this is on master, it will go into 1.19.0. We will probably never release 1.18.1, never mind porting this stuff there.

I assume the MWE is now working, does it? I should now test on PowerGrids, I'll do that ASAP.

**Comment by casella** [comment]
TODO:
- [ ] reduce -nlssMinSize default to 1000
- [ ] reduce -nlsMinSize default to 1000
- [ ] make sure that KLU is selected as linear solver if Kinsol is selected automatically
- [ ] retry with default dense solver if kinsol fails but size is less than nlssMinSize

**Comment by phannebohm** [comment]
> I assume the MWE is now working, does it? I should now test on PowerGrids, I'll do that ASAP.

It is not working. Kinsol is selected and tries 5 times in a row returning KIN_PRECONDSOLVE_FAILURE and then gives up.
If I manually set `-nlsLS=klu` the MWE finishes successfully, so this is what I will work on now. You can probably wait a bit with PowerGrids.

**Comment by casella** [comment]
Great! Let's see the outcome of the library testsuite.

**Comment by phannebohm** [comment]
Looks like #8136 fixed most of the regressions introduced in #7933, see the reports: [old](https://libraries.openmodelica.org/branches/history/master/2021-11-12%2022:41:02..2021-11-13%2019:36:18.html) and [new](https://libraries.openmodelica.org/branches/history/master/2021-11-15%2011:28:36..2021-11-15%2022:34:45.html).

I suppose the backup plan with dense solvers is still essential, but I think we are better off than before.

**Comment by casella** [comment]
Agreed. There are still a few regressions, but it's not nearly as bad as it was before these last commits.

We need to eventually get this fixed by the time we release 1.19.0.

In the meantime, we could probably get some more tests running again, e.g. [HelmholtzMedia.Examples.BranchingDynamicPipes.sim](https://libraries.openmodelica.org/branches/master/HelmholtzMedia/files/HelmholtzMedia_HelmholtzMedia.Examples.BranchingDynamicPipes.sim) by reducing the default of -nlssMaxDensity from 0.20 to 0.10, so that the dense solver is used anyway in those cases. I guess 0.10 is still large enough to prevent the use of the dense solver in cases where its performance would be really bad compared to the sparse one. Those have sizes well above 200 equations, and it's very rare to have more than 10% nonzero elements in each row of the Jacobian, i.e. 20 variables showing up in each equation, normally it's more like two to five.

@phannebohm would you like to give this a try?

Also, threre are several tests that started failing after your commit because of assertions triggered during simulation, e.g. [Buildings_3.0.0_Buildings.Fluid.MixingVolumes.Validation.MixingVolume](https://libraries.openmodelica.org/branches/master/Buildings_3.0.0/files/Buildings_3.0.0_Buildings.Fluid.MixingVolumes.Validation.MixingVolume.sim) or [ModelicaTest_3.2.1_ModelicaTest.MultiBody.FourbarVariants.JointUSR](https://libraries.openmodelica.org/branches/master/ModelicaTest_3.2.1/files/ModelicaTest_3.2.1_ModelicaTest.MultiBody.FourbarVariants.JointUSR.sim).

I have no idea why this happens. Could this be related to #6195?

 If you can't get this done in time, we could reduce the default of -nlssMaxDensity from 0.20 to 0.07 to make sure that those cases still run with the internal solver.

**Comment by casella** [comment]
@phannebohm, we also have problems on the large-scale gas network models that @matteodepascali is working on, because the sparse solver is now selected, given that the density is slightly less than 0.2. Those problems were unfortunately not solved when the sparse linear solver was introduced.

While we try to figure out what went wrong, I would really suggest to reduce the default of -nlssMaxDensity from 0.20 to 0.10 ASAP. This is a hack, but it will solve a lot of issues that would otherwise be very annoying to people using the nightly.

**Comment by AnHeuermann** [comment]
> Also, threre are several tests that started failing after your commit because of assertions triggered during simulation, e.g. Buildings_3.0.0_Buildings.Fluid.MixingVolumes.Validation.MixingVolume or ModelicaTest_3.2.1_ModelicaTest.MultiBody.FourbarVariants.JointUSR.
> 
> I have no idea why this happens. Could this be related to #6195?

Yes, at least if they had asserts before. https://github.com/OpenModelica/OpenModelica/commit/1cbaefaede641a04c749459ab8eca5bdc70a04d6 didn't introduce new asserts.

**Comment by phannebohm** [comment]
> In the meantime, we could probably get some more tests running again, e.g. [HelmholtzMedia.Examples.BranchingDynamicPipes.sim](https://libraries.openmodelica.org/branches/master/HelmholtzMedia/files/HelmholtzMedia_HelmholtzMedia.Examples.BranchingDynamicPipes.sim) by reducing the default of -nlssMaxDensity from 0.20 to 0.10, so that the dense solver is used anyway in those cases. I guess 0.10 is still large enough to prevent the use of the dense solver in cases where its performance would be really bad compared to the sparse one. Those have sizes well above 200 equations, and it's very rare to have more than 10% nonzero elements in each row of the Jacobian, i.e. 20 variables showing up in each equation, normally it's more like two to five.
> 
> @phannebohm would you like to give this a try?

See #8167. I will wait for the report and then close this issue for now. I would expect a few improvements and performance regressions.

@casella will open another ticket for the fallback strategy on the nls linear solver. This will have to wait for 1.20 I guess.

**Comment by casella** [comment]
@phannebohm, the Jenkings test jobs are broken again, I wrote @sjoelund about this, I hope it gets fixed soon. Unfortunately results will be mixed up with other stuff, I hope there won't be many regressions in there.

I'm going to close this ticket now, since I am pretty sure the 0.10 trick will work, all the tests that failed had densities above 0.10 and below 0.20. If it doesn't, we can reopen it.

**Comment by casella** [comment]
This also resolves #6342, I'm going to close that as well
